nemo/src/execution/planning/aggregates.rs

```diff
@@ -0,0 +1,212 @@
+};
+
+pub(super) fn generate_node_aggregate(
+    current_plan: &mut SubtableExecutionPlan,
+    mut variable_order: VariableOrder,
+    mut node: ExecutionNodeRef,
+    aggregates: &Vec<ChaseAggregate>,
+    aggregate_group_by_variables: &HashSet<Variable>,
+) -> (ExecutionNodeRef, VariableOrder) {
+    assert!(
+        aggregates.len() <= 1,
+        "currently only one aggregate term per rule is supported"
+    );
+    for aggregate in aggregates {
+        if reorder_required(&variable_order, aggregate_group_by_variables, aggregate) {
+            // Perform reordering of input trie scan and materialize into temporary trie
+
+            // Build variable order
+            // Start with group-by variables, followed by the aggregate input variables
+            // Try to preserve as much of the relative ordering as possible, to reduce the amount of swap that need to be done during the projecting/reordering
+            let input_variable_set = HashSet::from_iter(aggregate.input_variables.iter().cloned());
+            let mut variable_order_after_reordering =
+                variable_order.restrict_to(aggregate_group_by_variables);
+            for input_variable in variable_order
+                .restrict_to(&input_variable_set)
+                .as_ordered_list()
+            {
+                variable_order_after_reordering.push(input_variable);
+            }
+
+            let reordering_column_indices: Vec<_> = variable_order_after_reordering
+                .as_ordered_list()
+                .iter()
+                .map(|variable| *variable_order.get(variable).unwrap())
+                .collect();
+
+            let reordering = ProjectReordering::from_vector(reordering_column_indices, 3);
+
+            node = current_plan.plan_mut().project(node, reordering);
+
+            current_plan.add_temporary_table(node.clone(), "Subtable Aggregate Reorder");
+
+            // Update variable order to reordering
+            variable_order = variable_order_after_reordering;
+        }
+
+        let aggregated_column_index = *variable_order
+            .get(aggregate.aggregated_input_variable())
+            .expect("variable that is aggregated has to be in the variable order");
+
+        // Determine `last_distinct_column_index`
+        let mut last_distinct_column_index = 0;
+        for (index, variable) in variable_order.iter().enumerate() {
+            if aggregate.input_variables.contains(variable) {
+                last_distinct_column_index = index;
+            }
+        }
+
+        node = current_plan.plan_mut().aggregate(
+            node,
+            AggregationInstructions {
+                aggregate_operation: aggregate.aggregate_operation,
+                group_by_column_count: aggregate_group_by_variables.len(),
+                aggregated_column_index,
+                last_distinct_column_index,
+            },
+        );
+
+        // Update variable order after aggregation
+        {
```

Comment:
Why is this inside a block?

Reply:
I see. Not sure if this really helps, for me just having the lines below one another without any blank lines is grouping enough, but we can leave it this way for all I care.